File: problem/0704_Binary Search.js
// Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.

// You must write an algorithm with O(log n) runtime complexity.

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search  = function(nums, t) {
  return binary(nums, t, 0, nums.length-1)
};

function binary(arr, target, min, max){
  if (min > max) return -1
  let mid = parseInt((min+max)/2, 10)
  if (arr[mid] === target) return mid;
  if (target > arr[mid]) return binary(arr, target, mid+1, max)
  if (target < arr[mid]) return binary(arr, target, min, mid-1) 
}

File: c#-solution/0704. Binary Search.cs
// 0704. Binary Search
// https://leetcode.com/problems/binary-search/
// Tag: C#, Array, Binary Search, Easy

public class Solution {
    public int Search(int[] nums, int target) {
        int min = 0;
        int max = nums.Length -1;
        while(max >= min)
        {
            int mid = (max+min)/2;
            if(nums[mid] == target)
            {
                return mid;
            } else if(nums[mid] > target)
            {
                max = mid -1;
            } else 
            {
                min = mid +1;
            }
        }
        return -1;
    }
}